backends/simplego/dotgeneral_block.go

```diff
@@ -0,0 +1,323 @@
+//
+// Parameters:
+//   - node: the input node to potentially block
+//   - crossSize, contractingSize: normalized sizes for blocking calculation
+func shouldPreBlock(node *Node, crossSize, contractingSize int) bool {
+	dtype := node.shape.DType
+
+	// Check dtype is supported for blocking
+	switch dtype {
+	case dtypes.Float32, dtypes.Float64,
+		dtypes.Float16, dtypes.BFloat16,
+		dtypes.Int8, dtypes.Uint8,
+		dtypes.Int16, dtypes.Int32, dtypes.Int64,
+		dtypes.Uint16, dtypes.Uint32, dtypes.Uint64:
+		// dtype supported, continue to size check
+	default:
+		return false
+	}
+
+	// Must be large enough to benefit from blocking
+	blockDim := 1 << DotGeneralTargetBlockLog2Dim[dtype]
+	if crossSize < blockDim || contractingSize < blockDim {
+		return false
+	}
+
+	// Only pre-block constants or parameters (not computed values)
+	// This ensures we only pay the blocking cost once
+	switch node.opType {
+	case backends.OpTypeParameter, backends.OpTypeConstant:
+		return true
+	default:
+		return false
+	}
+}
+
+// shouldPreBlockRHS determines if the RHS (weights) should be pre-blocked for DotGeneral.
+// This is a specialized version for backward compatibility with 2D weight matrices.
+//
+// Pre-blocking is beneficial when:
+// - RHS is a 2D tensor [K, N] (standard weight layout)
+// - RHS has no batch dimensions (weights are shared across batch)
+// - Single contracting axis on axis 0 for RHS
+// - The dtype is supported for blocking
+// - The matrix is large enough to benefit from blocking (at least one full block in each dimension)
+func shouldPreBlockRHS(rhs *Node, lhsContractingAxes, rhsContractingAxes, rhsBatchAxes []int) bool {
+	// RHS (weights) must be 2D: [K, N]
+	if rhs.shape.Rank() != 2 {
+		return false
+	}
+
+	// RHS must have no batch dimensions (weights are shared across batch)
+	if len(rhsBatchAxes) != 0 {
+		return false
+	}
+
+	// Single contracting axis for RHS
+	if len(rhsContractingAxes) != 1 {
+		return false
+	}
+
+	// RHS contracts on axis 0 (standard [K, N] weight layout)
+	if rhsContractingAxes[0] != 0 {
+		return false
+	}
+
+	// LHS must also have single contracting axis
+	if len(lhsContractingAxes) != 1 {
+		return false
+	}
+
+	// Use the general shouldPreBlock with derived cross/contracting sizes
+	K := rhs.shape.Dimensions[0] // contracting size
+	N := rhs.shape.Dimensions[1] // cross size
+	return shouldPreBlock(rhs, N, K)
+}
+
+// getOrCreateBlockedInputGeneral returns a BlockForDotGeneral node for the given input tensor.
+// If a blocked version already exists (de-duplication), it returns the existing node.
+// Otherwise, it creates a new BlockForDotGeneral node.
+//
+// This is the generalized version that works for both LHS and RHS operands with any shape.
+//
+// Parameters:
+//   - input: the node to block
+//   - contractingAxes, batchAxes: axes from the original tensor shape
+//   - batchSize, crossSize, contractingSize: normalized sizes
+func (b *Builder) getOrCreateBlockedInputGeneral(input *Node,
```

Comment:
The "getOrCreate" prefix is not needed. With your "dedup" all ops are "getOrCreate"

Why not name it for what it does, `blockForDotGeneral()`, like the OpType ?

Reply:
Ugh, I just realized: the normal ops are not making use of the dedup code? Am I right ? 

I'm noticing none of the dedup tests actually test that with make the same `Add` operation twice, it actually only adds one node in the graph ... Sry, I should have caught that, we should fix that.

Reply:
Done in PR #172  :)

Reply:
That makes sense- I renamed it to `blockForDotGeneral`! And thanks for catching that; I merged in `main` so we're now using the `getOrCreateNode` throughout